### DISCUSSION TITLE
How to configure formattingProvider in the Emacs eglot package

### DISCUSSION THREAD

**Discussion by wolverian** [haskell/haskell-language-server#3306] [Q&A] [answered]
I'm not sure how to configure the `haskell.formattingProvider` setting in Emacs eglot. I see HLS has documentation for lap-mode instead, which is an alternative LSP client for Emacs. It would be great to add eglot instructions as well, as it is now merged to Emacs upstream and will be in the next release as a built-in LSP client.

I've tried setting `eglot-workspace-configuration` so that calling `eglot-show-workspace-configuration` prints:

```json
{
  "haskell.formattingProvider": "fourmolu"
}
```

If I run `M-x eglot-signal-didChangeConfiguration` I can see eglot sending a message to HLS:

```elisp
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:haskell\.formattingProvider "fourmolu")))
```

but I'm not seeing any kind of reaction from the server. There's nothing in the server's stderr either.

**Comment by wolverian** [comment]
I just realised that 1.8.0.0 doesn't support formatting with the fourmolu with GHC 9.4.2, which I was using. This might have been my fault then, sorry!

**Comment by michaelpj** [comment]
Re: eglot in general, someone has to write the docs! We need an eglot user who actually understands how this works to step up and write some documentation on how to do it. Maybe that's you :)

**Comment by unhammer** [comment, marked as answer]
Reading https://www.gnu.org/software/emacs/manual/html_node/eglot/Project_002dspecific-configuration.html#Examples it seems the correct way is this:

```emacs-lisp
(setq-default eglot-workspace-configuration '(:haskell (:formattingProvider "fourmolu")))
```
(or `M-x add-dir-local-variable` to make it project-specific). Then running `M-x eglot-show-workspace-configuration` I see 

```json
{
  "haskell": {
    "formattingProvider": "fourmolu"
  }
}
```

I'm not sure how to introspect what HLS actually parsed it as, but after setting that I get four spaces and leading commas in lists, whereas without it I got two spaces and trailing commas.

**Comment by kinthaiofficial** [comment]
Configuring the formatting provider in HLS via eglot is a bit more involved than most LSP servers because HLS has multiple formatters and eglot needs to know which one to request.

**eglot's formatting setup**:
eglot uses `lsp-format-document` / `eglot-format` which sends `textDocument/formatting` to the server. HLS will use whichever formatter is configured in its settings.

**Configuring HLS formatter via eglot**:
```elisp
(setq-default eglot-workspace-configuration
  '(:haskell
    (:formattingProvider "fourmolu"  ; or "ormolu", "stylish-haskell", "floskell", "brittany"
     :plugin
     (:fourmolu
      (:globalOn t)))))
```

Place this in your `init.el` before eglot connects, or use `dir-locals.el` per project.

**Verifying the configuration reached HLS**:
In HLS logs (`M-x eglot-stderr-buffer`), you should see the workspace configuration being applied on startup. If formatting still uses the wrong provider, the config may not be reaching HLS.

**Common issue: fourmolu not found**:
HLS looks for the `fourmolu` binary in PATH. If it's not installed, HLS silently falls back to another formatter. Install with:
```bash
cabal install fourmolu
# or
ghcup install fourmolu
```

**Per-project override**:
Use a `.hlsfp.yaml` or `fourmolu.yaml` in the project root for formatter-specific settings (line length, indentation style, etc.).

**Format-on-save**:
```elisp
(add-hook 'haskell-mode-hook
  (lambda ()
    (add-hook 'before-save-hook #'eglot-format nil t)))
```

Is the issue that HLS isn't formatting at all, or that it's using the wrong formatter?
